manta-pay/src/crypto/poseidon/hash.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Poseidon Hash Implementation
+
+use crate::crypto::poseidon::{Field, FieldGeneration, Permutation, Specification};
+use core::{fmt::Debug, hash::Hash};
+use manta_crypto::{
+    hash::ArrayHashFunction,
+    rand::{Rand, RngCore, Sample},
+};
+use manta_util::{
+    codec::{Decode, DecodeError, Encode, Read, Write},
+    vec::VecExt,
+};
+
+#[cfg(feature = "serde")]
+use manta_util::serde::{Deserialize, Serialize};
+
+/// Poseidon Hasher
+#[cfg_attr(
+    feature = "serde",
+    derive(Deserialize, Serialize),
+    serde(
+        bound(
+            deserialize = "Permutation<S, COM>: Deserialize<'de>, S::Field: Deserialize<'de>",
+            serialize = "Permutation<S, COM>: Serialize, S::Field: Serialize"
+        ),
+        crate = "manta_util::serde",
+        deny_unknown_fields
+    )
+)]
+#[derive(derivative::Derivative)]
+#[derivative(
+    Clone(bound = "Permutation<S, COM>: Clone, S::Field: Clone"),
+    Debug(bound = "Permutation<S, COM>: Debug, S::Field: Debug"),
+    Eq(bound = "Permutation<S, COM>: Eq, S::Field: Eq"),
+    Hash(bound = "Permutation<S, COM>: Hash, S::Field: Hash"),
+    PartialEq(bound = "Permutation<S, COM>: PartialEq, S::Field: PartialEq")
+)]
+pub struct Hasher<S, const ARITY: usize, COM = ()>
+where
+    S: Specification<COM>,
+{
+    /// Poseidon Permutation
+    permutation: Permutation<S, COM>,
+
+    /// Domain Tag
+    domain_tag: S::Field,
```

Comment:
`domain_tag` could be `S::ParameterField` according to our discussion with Dimitry.